### ISSUE TITLE
can I set content-type for requests?

### ISSUE THREAD

**Issue by yahyaKacem** [sahat/satellizer#479]
Hi, I need to make the login request get sent with a content-type: "application/x-www-form-urlencoded", also probably using the new [$httpParamSerializer](https://code.angularjs.org/1.4.2/docs/api/ng/service/$httpParamSerializer) and [$httpParamSerializerJQLike](https://code.angularjs.org/1.4.2/docs/api/ng/service/$httpParamSerializerJQLike) from AngularJS 1.4.2, I checked the source code your using the $http.post without a configuration object so no possibility for request configuration there, so is there any plans for adding support to configuring the $http request sent by satellizer?
just to be more specific here's what I want to do:
I want to make a $http request for login with this property:

    $http.post({
      url: "login_url",
      method: "POST",
      data: $httpParamSerializerJQLike(user),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'}
    });

**Comment by sahat** [comment]
@yahyaKacem are you suggesting an addition to this call?
```js
$auth.authenticate(name, [userData, requestOptions])
```

**Comment by yahyaKacem** [comment]
sorry, I wasn't clear, I'm suggestion an addition to the regular login/signup without provider, for this calls:

    $auth.login(user, redirect);
    $auth.signup(user);
so in this 2 functions [local.login](https://github.com/sahat/satellizer/blob/master/satellizer.js#L423), [local.signup](https://github.com/sahat/satellizer/blob/master/satellizer.js#L432) instead of calling the $http.post short-cut use the fully configurable $http(config) format and change the call to be configurable like this:

    $auth.login(user, redirect, requestConfig);
    $auth.signup(user, requestConfig);
where requestConfig is an object with additional configuration options like:

    {
      url: "login_url",
      data: $httpParamSerializerJQLike(user),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'}
    }

Also the local login and signup should be adjusted to have a basic configuration like method and stuff.
for now I'm having this login decorator I'm using to decorate the `$auth.login` function:

    $delegate.login = function loginF(user, redirect, requestConfig) {
      var handleLoginResponse;
      requestConfig         = requestConfig || {};
      requestConfig.method  = "POST";
      requestConfig.data    = $httpParamSerializerJQLike(user);
      requestConfig.headers = {"Content-Type": "application/x-www-form-urlencoded"};
      requestConfig.url     = config.baseUrl ? utils.joinUrl(config.baseUrl, config.loginUrl) : config.loginUrl;
      handleLoginResponse   = function handleLoginResponseF(response) {
        shared.setToken(response, redirect);
        return response;
      };
      return $http(requestConfig).then(handleLoginResponse);
    };

also I'm going to make something like that for $auth.signup too.
I hope this showcase what I meant exactly.

**Comment by sahat** [comment]
Are there any downsides to using `$httpParamSerializerJQLike` by default or should it remain optional?

I could modify the function as follows:

```js
local.login = function(user, options) {
          var loginUrl = config.baseUrl ? utils.joinUrl(config.baseUrl, config.loginUrl) : config.loginUrl;
          options = options || {};
          
          return $http({
            url: options.url || loginUrl,
            method: 'POST',
            data: options.paramSerializer ? $httpParamSerializerJQLike(user) : user,
            headers: options.headers
          }).then(function(response) {
            shared.setToken(response, options);
            return response;
          });
        };
```

Then in your controller:

```js
$scope.login = function() {
      var loginOptions = {
        paramSerializer: true,
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
      };
      $auth.login({ email: $scope.email, password: $scope.password }, loginOptions)
        .then(function() {
          // done
        })
```

**Comment by yahyaKacem** [comment]
I think it should remain optional users may want to use their own serializers also as a suggestion:

    local.login = function(user, options) {
      var loginUrl, defaultOptions, requestDefinitionObject;
      loginUrl = config.baseUrl ? utils.joinUrl(config.baseUrl, config.loginUrl) : config.loginUrl;
      defaultOptions = {
        data: user,
        url: loginUrl,
        method: 'POST'
      };
      requestDefinitionObject = angular.extend(defaultOptions, options);
      return $http(requestDefinitionObject).then(function(response) {
        shared.setToken(response, options);
        return response;
      });
    };

then in the controller:

    $scope.login = function() {
      var loginOptions;
      loginOptions = {
        paramSerializer: '$httpParamSerializerJQLike',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
      };
      // or
      loginOptions = {
        paramSerializer: '$httpParamSerializer',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
      };
      // or
      loginOptions = {
        paramSerializer: 'customParamSerializerDefinedAsService',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
      };
      // or
      loginOptions = {
        paramSerializer: function cutomParamsSerializerDefinedInline() {/*serialize*/},
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
      };
      $auth.login({ email: $scope.email, password: $scope.password }, loginOptions)
           .then(function() {
             // done
           });
    }

Check the new paramSerializer argument for the [$http](https://code.angularjs.org/1.4.2/docs/api/ng/service/$http) service.

**Comment by sahat** [comment]
@yahyaKacem Thanks for the code example. Do you know by any chance if the following are equivalent when making a POST request?

```js
paramSerializer: '$httpParamSerializerJQLike'
```

vs

```js
    data: $httpParamSerializerJQLike(myData),
```

I'll double check again, but last time I tried the first approach did not work for me, while the second approach did work.

**Comment by yahyaKacem** [comment]
you'r right this `paramSerializer: '$httpParamSerializerJQLike'` doesn't work for the POST requests, so a solution could be like this the first part stays the same:

    local.login = function(user, options) {
      var loginUrl, defaultOptions, requestDefinitionObject;
      loginUrl = config.baseUrl ? utils.joinUrl(config.baseUrl, config.loginUrl) : config.loginUrl;
      defaultOptions = {
        data: user,
        url: loginUrl,
        method: 'POST'
      };
      requestDefinitionObject = angular.extend(defaultOptions, options);
      return $http(requestDefinitionObject).then(function(response) {
        shared.setToken(response, options);
        return response;
      });
    };
but in the controller it will become:

    $scope.login = function() {
      var user, loginOptions, customParamSerializerDefinedInController;
      customParamSerializerDefinedInController = function customParamSerializerDefinedInController(data) {
        // return custom serialized data
      };
      user = $httpParamSerializer({email: $scope.email, password: $scope.password});
      // or
      user = customParamSerializerDefinedInController({email: $scope.email, password: $scope.password});
      // or
      user = customParamSerializerDefinedAsService({email: $scope.email, password: $scope.password});
      // or
      user = $httpParamSerializerJQLike({email: $scope.email, password: $scope.password});
      loginOptions = {
        headers: {'Content-Type': 'application/x-www-form-urlencoded'}
      };
      $auth.login(user, loginOptions)
           .then(function () {
             // done
           });
    }

**Comment by iammary** [comment]
@sahat when can this be added to the release? :+1:

**Comment by sahat** [comment]
@iammary Hopefully this weekend, however, there will be breaking changes in 0.12, just so you know.

**Comment by sahat** [comment]
Resolved in [0.12](https://github.com/sahat/satellizer/releases/tag/0.12.0) release.

**Comment by iammary** [comment]
@sahat thanks a lot! :+1: :+1:
